### PR TITLE
#13249: Update doc strings

### LINKED ISSUE #13249: Update CCL op docstrings
Our docs generation for ops has changed recently and CCL ops should be updated to define their doc strings in a compatible way.

> you don't need to spell out the function signature explicitly as the first line of the docstring anymore. The signature is now parsed from the args, keywords args and return sections of the docstring. Another change is how those sections are formatted. We now use the Google style docstrings, which is the most common style (used by pytorch, for instance). Here's a full example of the docstring for the embedding op. Notice how we specify if the argument is optional, and how we provide the default value if there is one.
When you add a new op, you only need to list it in the api.rst file, and no additional files need to be created.
        Retrieves word embeddings using input_tensor. The input_tensor is a list of indices, and the embedding matrix, and the output is the corresponding word embeddings.

> ```
>         Args:
>             input_tensor (ttnn.Tensor): the input indices tensor.
>             weight (ttnn.Tensor): the embeddings tensor that corresponds to the indices tensor.
> 
> 
>         Keyword args:
>             memory_config (ttnn.MemoryConfig, optional): Memory configuration for the operation. Defaults to `input tensor memory config`.
>             output_tensor (ttnn.Tensor, optional): Preallocated output tensor. Defaults to `None`.
>             queue_id (int, optional): command queue id. Defaults to `0`.
>             padding_idx (int, optional): the padding token. Default to `None`.
>             layout (ttnn.Layout): the layout of the output tensor. Defaults to `ttnn.ROW_MAJOR_LAYOUT`.
>             embeddings_type (ttnn.EmbeddingsType): the type of embeddings. Defaults to `ttnn._ttnn.operations.embedding.EmbeddingsType.GENERIC`.
>             dtype (ttnn.DataType, optional): the data type for the output tensor. Defaults to `None`.
> 
> 
>         Returns:
>             ttnn.Tensor: the output tensor.
> 
> 
>         Example:
>             >>> device_id = 0
>             >>> device = ttnn.open_device(device_id=device_id)
>             >>> tensor = ttnn.to_device(ttnn.from_torch(torch.tensor([[1, 2, 4, 5], [4, 3, 2, 9]]), dtype=ttnn.uint32), device=device)
>             >>> # an embedding matrix containing 10 tensors of size 4
>             >>> weight = ttnn.to_device(ttnn.from_torch(torch.rand(10, 4), dtype=ttnn.bfloat16), device=device)
>             >>> output = ttnn.embedding(tensor, weight)
>             ttnn.Tensor([ [[1, 0.106445, 0.988281, 0.59375],
>                 [0.212891, 0.964844, 0.199219, 0.996094],
>                 [3.78362e-38, 0, 7.89785e-39, 0],
>                 [8.04479e-38, 0, 1.25815e-38, 0]],
>             [[2.71833e-38, 0, 3.59995e-38, 0],
>                 [7.60398e-38, 0, 1.83671e-38, 0],
>                 [2.22242e-38, 0, 1.88263e-38, 0],
>                 [1.35917e-38, 0, 4.49994e-39, 0]]], dtype=bfloat16))
> ```

Additionally, we should update/improve our examples to reflect the overloads and to mirror the new mesh tensor/device mesh APIs

### PR DESCRIPTION
**Pull request by Aswinmcw**
Tracking Issue : #13249 

Changes Done:

- Modified docs to newer format
- Modified README

Doc Preview:
Screenshot 2024-10-01 at 3 44 37 PM

Screenshot 2024-10-03 at 8 52 52 PM

Screenshot 2024-10-03 at 8 52 29 PM

### Checklist
- [x] Post commit CI passes - https://github.com/tenstorrent/tt-metal/actions/runs/11124510537
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

**Review by yieldthought — APPROVED** [review]
Looks like this has picked up the all_gather/line_all_gather work by mistake?

**Comment by SeanNijjar** [comment]
> Looks like this has picked up the all_gather/line_all_gather work by mistake?

I think so. Maybe the PR should be reopened to after rebase and conflict resolution to minimize spam.

**Comment by Aswinmcw** [comment]
> > Looks like this has picked up the all_gather/line_all_gather work by mistake?
> 
> I think so. Maybe the PR should be reopened to after rebase and conflict resolution to minimize spam.

Rebased!

**Comment by SeanNijjar** [comment]
Can you please update the examples to represent a real usage? We don't run all-gather (or reduce-scatter) on single device. I think two examples would be useful: one with ShardTensor2dMesh (see `test_all_gather_TG_post_commit.py` for reference) and another with the aggregate_as_tensor (as in `test_all_gather.py`).

File: ttnn/cpp/ttnn/operations/ccl/all_gather/all_gather_pybind.cpp [hunk 1]
```diff
@@ -78,33 +78,42 @@ void py_bind_all_gather(pybind11::module& module) {
...
+            topology (ttnn.Topology, optional): The topology configuration to run the operation in. Valid options are Ring and Linear. Defaults to `ttnn.Topology.Ring`.
 
-        Example:
+        Returns:
+            ttnn.Tensor: the output tensor.
 
-            >>> tensor = ttnn.from_torch(torch.tensor((1, 2), dtype=torch.bfloat16), device=device)
-            >>> output = ttnn.all_gather(tensor, dim=0)
+        Example:
+            >>> full_tensor = torch.randn([1, 1, 32, 32], dtype=torch.bfloat16)
```
Comment (SeanNijjar) [inline ttnn/cpp/ttnn/operations/ccl/all_gather/all_gather_pybind.cpp:105]: I believe the inner dim here should be 32 * num_chips (=> 256)

File: ttnn/cpp/ttnn/operations/ccl/reduce_scatter/reduce_scatter_pybind.cpp [hunk 2]
```diff
@@ -55,25 +55,38 @@ void py_bind_reduce_scatter(pybind11::module& module) {
...
+            topology (ttnn.Topology, optional): The topology configuration to run the operation in. Valid options are Ring and Linear. Defaults to `ttnn.Topology.Ring`.
+
+        Returns:
+            ttnn.Tensor: the output tensor.
 
         Example:
 
-            >>> tensor = ttnn.from_torch(torch.tensor((1, 2), dtype=torch.bfloat16), device=device)
-            >>> output = ttnn.reduce_scatter(tensor, dim=0)
+            >>> full_tensor = torch.randn([1, 1, 32, 32], dtype=torch.bfloat16)
```
Comment (SeanNijjar) [inline ttnn/cpp/ttnn/operations/ccl/reduce_scatter/reduce_scatter_pybind.cpp:78]: The same comment applies as above. but needs to be * 8 * 8 because each input tensor must be fractured 8 ways (num chips) and then we need 8 of those for the mesh tensor (which will fracture by 8 to get the per chip shape under the hood).

**Review by SeanNijjar — COMMENTED** [review]
Hey @Aswinmcw - will these examples will work? They look different from how we run in the tests (specifically the input tensor shape).

**Review by SeanNijjar — APPROVED** [review]

**Review by yan-zaretskiy — APPROVED** [review]
